src/representation_theory/basic.lean

```diff
@@ -107,4 +128,121 @@ theorem char_conj (g : G) (h : G) : (character k G V) (h * g * h⁻¹) = (charac
 
 end character
 
+section linear_hom
+
+variables {k G V W : Type*} [comm_semiring k] [group G] [add_comm_monoid V] [add_comm_monoid W]
+variables [module k V] [distrib_mul_action G V] [smul_comm_class G k V]
+variables [module k W] [distrib_mul_action G W] [smul_comm_class G k W]
+
+/--
+Given two `k`-linear representations `V` and `W` of a group `G`,
+the space are linear maps `V →ₗ[k] W` is itself a representation of `G`.
+-/
+instance  linear_hom_distrib_mul_action : distrib_mul_action G (V →ₗ[k] W) :=
+{ smul := λ g f, ↑(as_group_hom k G W g) ∘ₗ f ∘ₗ ↑(as_group_hom k G V g)⁻¹,
+  one_smul := λ f, by {ext, simp only [map_one, units.coe_one, one_inv, coe_comp, function.comp_app,
+                      one_apply]},
+  mul_smul := λ g h f, by {ext, simp},
+  smul_add := λ g f₁ f₂, by simp only [add_comp, comp_add],
+  smul_zero := λ g, by simp only [linear_map.zero_comp, linear_map.comp_zero]}
+
+@[simp]
+lemma linear_hom_smul_def (g : G) (f : V →ₗ[k] W) :
+  g • f = ↑(as_group_hom k G W g) ∘ₗ f ∘ₗ ↑(as_group_hom k G V g)⁻¹ := rfl
+
+instance linear_hom_smul_comm_class : smul_comm_class G k (V →ₗ[k] W) :=
+{smul_comm := λ g r f, by {ext, simp only [linear_hom_smul_def, coe_comp, function.comp_app,
+                          smul_apply, map_smulₛₗ, ring_hom.id_apply]}}
+
+instance dual_distrib_mul_action : distrib_mul_action G (module.dual k V)  :=
+{ smul := λ g f, f ∘ₗ ↑(as_group_hom k G V g)⁻¹,
+  one_smul := λ f, by {ext, simp only [map_one, one_inv, units.coe_one, coe_comp, function.comp_app,
+                      one_apply]},
+  mul_smul := λ g h f, by {ext, simp only [map_mul, mul_inv_rev, units.coe_mul, coe_comp,
+                          function.comp_app, linear_map.mul_apply]},
+  smul_add := λ g f₁ f₂, by simp only [add_comp],
+  smul_zero := λ g, by simp only [zero_comp]}
+
+@[simp]
+lemma dual_smul_def (g : G) (f : module.dual k V) : g • f = f ∘ₗ ↑(as_group_hom k G V g)⁻¹ := rfl
+
+instance dual_smul_comm_class : smul_comm_class G k (module.dual k V) :=
+{smul_comm := λ g r f, by {ext, simp only [dual_smul_def, coe_comp, function.comp_app, smul_apply]}}
+
+end linear_hom
+
+section tensor
+
+open_locale tensor_product
+
+variables (k G V W : Type*) [comm_semiring k] [monoid G] [add_comm_monoid V] [add_comm_monoid W]
+variables [module k V] [distrib_mul_action G V] [smul_comm_class G k V]
+variables [module k W] [distrib_mul_action G W] [smul_comm_class G k W]
+
+/--
+Given two `k`-linear representations `V` and `W`, their tensor product over `k` is itself a
+representation, characterized by `g • (v ⊗ₜ w) = (g • v) ⊗ₜ (g • w)` for `g : G`, `v : V`, `w : W`.
+-/
+@[derive [add_comm_monoid, module k, inhabited]]
+def tensor_rep : Type* := (V ⊗[k] W)
+
+notation V ` ⊗[` k `,` G `] ` W := tensor_rep k V W
+
+variables {k G V W}
+
+/--
+The tensor product representation `V ⊗[k,G] W` is equivalent to `V ⊗[k] W` as a `k`-module
+-/
+@[simp]
```

Comment:
```suggestion
```
These should not be marked simp

Reply:
Sure, I added the simp tags because implementing your comment about `tensor_rep_def` made a lot of `to_tensor_rep` and `of_tensor_rep` appear in subsequent proofs, and I needed to simplify those away to make things work as before. I guess I can just do `simp [to_tensor_rep, of_tensor_rep]` wherever that happens. May I know what's the reason for not marking these as simp, by the way?

Reply:
Have a look at `multiplicative.of_add` and friends for inspiration. The idea is that you never unfold these functions, and instead push them around until you can cancel them out as a pair. That means you can always tell by looking at the goal state which parts refer to the synonym and which parts refer to the original type.